### PR TITLE
[Spark-40099][SQL] Merge adjacent CaseWhen branches if their values are the same

### PR DESCRIPTION
**Pull request by Le-Dai**
### What changes were proposed in this pull request?

Merge adjacent CaseWhen branches if their values are the same

### Why are the changes needed?

Improvement of Optimized Logical Plan

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

add unit test

### CODE REVIEW

**Comment by MrDLontheway** [comment]
@wangyum 
pls help review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -639,6 +639,26 @@ object RemoveNoopOperators extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Merge case condition expression with same value.
+ */
+object MergeConditionWithValue extends Rule[LogicalPlan] {
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:645]: Do you have this rule execution time if it has thousands of branches.

Comment (MrDLontheway) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:645]: this is rule execution time for 2000 branches 
image

Comment (MrDLontheway) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:645]: @wangyum  have any update ?

**Pull request closed without merging**